module/documents/item.mjs

```diff
@@ -816,23 +813,26 @@ export default class Item5e extends Item {
     if ( Hooks.call("dnd5e.preUseItem", item, config, options) === false ) return;
 
     // Display configuration dialog
-    if ( (options.configureDialog !== false) && config.needsConfiguration ) {
-      const configuration = await AbilityUseDialog.create(item);
+    if ( (options.configureDialog !== false) && needsConfiguration ) {
+      const configuration = await AbilityUseDialog.create(item, config);
       if ( !configuration ) return;
       foundry.utils.mergeObject(config, configuration);
     }
 
-    // Handle spell upcasting
-    if ( isSpell && (config.consumeSpellSlot || config.consumeSpellLevel) ) {
-      const upcastLevel = config.consumeSpellLevel === "pact" ? as.spells.pact.level
-        : parseInt(config.consumeSpellLevel);
-      if ( upcastLevel && (upcastLevel !== is.level) ) {
-        item = item.clone({"system.level": upcastLevel}, {keepId: true});
+    // Handle upcasting
+    if ( item.type === "spell" ) {
+      let level = null;
+      if ( config.slotLevel ) {
+        // A spell slot was consumed.
+        level = config.slotLevel === "pact" ? as.spells.pact.level : parseInt(config.slotLevel.at(-1));
```

Comment:
I'd be happier with a `config.slotLevel.replace("spell", "")` in that case and gain some uniformity for the inputs, all being strings.

Reply:
Yeah, that makes sense, since we accept the `"pact"` string as an option. I think just for ergonomics it's pretty easy for us to support `number` though as well, e.g.:
```js
Number.isInteger(config.slotLevel) ? config.slotLevel : config.slotLevel === "pact" ? as.spells.pact.level : parseInt(config.slotLevel.replace("spell", ""));
```
It's a bit better for backwards-compatibility too.

Reply:
> Yeah, that makes sense, since we accept the `"pact"` string as an option. I think just for ergonomics it's pretty easy for us to support `number` though as well, e.g.:
> 
> ```js
> Number.isInteger(config.slotLevel) ? config.slotLevel : config.slotLevel === "pact" ? as.spells.pact.level : parseInt(config.slotLevel.replace("spell", ""));
> ```
> 
> It's a bit better for backwards-compatibility too.

Since we're going for zero backwards compatibility already (`false` in the config will not be doing what it does now), I don't see much point in going half-way for this.

Unless I am mistaken, will we not at one point (unless we do so already) support custom spell progressions beyond `spellx` and `pact`? If so, seems like a string key is a one-size-fits-all.

Reply:
> Unless I am mistaken, will we not at one point (unless we do so already) support custom spell progressions beyond `spellx` and `pact`? If so, seems like a string key is a one-size-fits-all.

We transact spell slot levels throughout the codebase as numbers. It's not difficult to imagine our own code or 3rd-party code having the spell level as a number, but when passing it to this method they have to awkwardly cast it to a string with `level.toString()`. I think a one-size-fits-all solution should really allow for numbers.

Reply:
Not married to either, so sure, not an issue.